add octo sts tutorials (#2861)
chainguard-dev/edu: content/open-source/octo-sts/trust-policy-reference.md

description: "Learn about Octo STS trust policy configuration including claim matching, permission settings, pattern validation, and advanced use cases"
date: 2025-12-22T15:04:05+01:00
lastmod: 2025-12-22T15:04:05+01:00
tags: ["octo-sts", "Reference", "Configuration"]
draft: false
images: []
menu:
  docs:
    parent: "octo-sts"
weight: 50
toc: true
---

Trust policies are YAML configuration files that control how Octo STS validates OIDC tokens and issues GitHub access tokens. This reference provides documentation of trust policy syntax and options.

## File Location

Trust policies must be stored in your repository at:

```
.github/chainguard/{name}.sts.yaml
```

The `{name}` portion becomes the identity parameter in the token exchange request:

```
https://octo-sts.dev/sts/exchange?scope=org/repo&identity={name}
```

For example, `.github/chainguard/renovate.sts.yaml` would be referenced as `identity=renovate`.

## Policy Types

Octo STS supports two types of trust policies:

**Repository Trust Policy**: Grants access to a single repository. The policy file must be in the repository it grants access to.

> 

Reviewer: ```suggestion
```

> **Organization Trust Policy**: Grants access to multiple repositories within an organization. Uses additional `repositories` field to specify scope.

Reviewer: ```suggestion
* **Organization Trust Policy**: Grants access to multiple repositories within an organization. Uses additional `repositories` field to specify scope.
```

- Read: View discussions
- Write: Create and manage discussions

**environments** (`read`, `write`)
- Read: View environment configuration
- Write: Configure environments

**issues** (`read`, `write`)
- Read: View issues
- Write: Create, update, close issues

**packages** (`read`, `write`)
- Read: Download packages
- Write: Publish and delete packages

**pages** (`read`, `write`)
- Read: View GitHub Pages settings
- Write: Configure GitHub Pages

**projects** (`read`, `write`, `admin`)
- Read: View projects
- Write: Edit projects
- Admin: Manage project settings

**pull_requests** (`read`, `write`)
- Read: View pull requests
- Write: Create and manage pull requests

**statuses** (`read`, `write`)
- Read: View commit statuses
- Write: Create commit statuses

**workflows** (`read`, `write`)
- Read: View workflow files
- Write: Update workflow files (critical for tools like Renovate)

### Organization Permissions

**members** (`read`, `write`)
- Read: View organization members
- Write: Manage organization membership

**administration** (`read`, `write`)
- Read: View organization settings
- Write: Manage organization settings

### Additional Permissions

> For the complete and current list of permissions, see the [octo-sts README](https://github.com/octo-sts/app#octo-sts-github-permissions).

Reviewer: ```suggestion
For the complete and current list of permissions, Refer to the [octo-sts README](https://github.com/octo-sts/app#octo-sts-github-permissions).
```

permissions:
  contents: read
```

### Example 3: Kubernetes Service Account in Production Namespace

```yaml
issuer: https://kubernetes.default.svc.cluster.local
subject_pattern: "system:serviceaccount:production:.*"

permissions:
  contents: write
  deployments: write
```

### Example 4: GitHub Actions from Any Repository in Organization

```yaml
issuer: https://token.actions.githubusercontent.com
subject_pattern: "repo:chainguard-dev/.*:ref:refs/heads/main"
claim_pattern:
  repository_owner: "chainguard-dev"

permissions:
  contents: read
```

## Security Considerations

### Principle of Least Privilege

Grant only the minimum permissions required:

```yaml
# Good: Specific permissions for specific task
permissions:
  contents: read
  pull_requests: write
```

```yaml
# Avoid: Overly broad permissions
permissions:
  contents: write
  issues: write
  pull_requests: write
  workflows: write
  packages: write
```

### Subject Matching

Prefer exact subject matches over patterns:

```yaml
# Better: Exact match
subject: repo:org/repo:ref:refs/heads/main
```

```yaml
# Acceptable but broader: Pattern match
subject_pattern: "repo:org/repo:ref:refs/heads/.*"
```

### Branch Protection

> Even with `contents: write`, enable branch protection to require pull requests:

Reviewer: This line doesn't scan for me. Is it saying "even with this permission, you should still require it to submit pull requests rather than merging changes without review"?
